docs/architecture/adr-047-handling-evidence-from-light-client.md

```diff
@@ -0,0 +1,177 @@
+# ADR 047: Handling evidence from light client
+
+## Changelog
+* 18-02-2020: Initial draft
+* 24-02-2020: Second version
+
+## Context
+
+If the light client is under attack (either directly - lunatic / phantom
+validators or indirectly - fork on the main chain), it's supposed to halt and
+send evidence of misbehavior to the correct full node. Upon receiving an
+evidence, full node should punish malicious validators (if possible).
+
+## Decision
+
+When a light client sees two conflicting headers (`H1.Hash() != H2.Hash()`,
+`H1.Height == H2.Height`), both having 1/3+ of voting power of the currently
+trusted validator set, it will submit a `ConflictingHeadersEvidence` to all
+full nodes it's connected to.
+
+```go
+type ConflictingHeadersEvidence struct {
+  H1 types.SignedHeader
+  H2 types.SignedHeader
+}
+```
+
+When a full node receives the `ConflictingHeadersEvidence` evidence, it should
+a) validate it b) figure out if malicious behaviour is obvious (immediately
+slashable) or the fork accountability protocol needs to be started.
+
+### Validating headers
+
+Check both headers are valid (`ValidateBasic`), have
+the same height and signed by 1/3+ of known validator set.
+
+- Q: What if light client validator set is not equal to full node's validator
+  set (i.e. from full node's point of view both headers are not properly signed;
+  this includes the case where none of the two headers were committed on the
+  main chain)
+
+  Reject the evidence. It means light client is following a fork, but, hey, at
+  least it will halt.
+
+- Q: Don't we want to punish validators who signed something else even if they
+  have less or equal than 1/3?
+
+  No consensus so far. Ethan said no, Zarko said yes.
+  https://github.com/tendermint/spec/pull/71#discussion_r374210533
+
+### Figuring out if malicious behaviour is immediately slashable
+
+Let's say H1 was committed on the main chain. Intersect validator sets of H1
```

Comment:
What is "main chain"? Will a full node know that H1 was committed on the "main" chain? The full node sees two headers at same height, H1 and H2. It could check its state and see that it has one, let's say H1. H2 may indicate either a full-fork or a light-fork but I don't think the full node can distinguish. 
Maybe "the full node's chain" or something along those lines.

Reply:
> What is "main chain"

the longest branch in Bitcoin, the branch 2/3+ of vals voted for in Tendermint (note: can be a fork if >= 1/3+ are malicious).

> Will a full node know that H1 was committed on the "main" chain?

if there is an actual fork (chain split), full node may follow either one or another branch. It's supposed to halt if it notices the full fork. So both H1 or H2 can be considered committed on the main chain depending on which branch the full node is following.

> H2 may indicate either a full-fork or a light-fork but I don't think the full node can distinguish.

It can because it knows more than the light client (app state for example).